crates/matrix-sdk/src/event_cache/deduplicator.rs

```diff
@@ -0,0 +1,266 @@
+            let Some(event_id) = event.event_id() else {
+                // The event has no `event_id`.
+                return Decoration::Invalid(event);
+            };
+
+            if self.bloom_filter.lock().unwrap().check_and_set(&event_id) {
+                // Bloom filter has false positives. We are NOT sure the event is NOT present.
+                // Even if the false positive rate is low, we need to iterate over all events to
+                // ensure it isn't present.
+
+                // But first, let's ensure `event` is not a duplicate from `events_to_scan`,
+                // i.e. if the iterator itself contains duplicated events! We use a `BTreeSet`,
+                // otherwise using a bloom filter again may generate false positives.
+                if already_seen.contains(&event_id) {
+                    // The iterator contains a duplicated `event`.
+                    return Decoration::Duplicated(event);
+                }
+
+                // Now we can iterate over all events to ensure `event` is not present in
+                // `existing_events`.
+                let duplicated = existing_events.revents().any(|(_position, other_event)| {
+                    other_event.event_id().as_ref() == Some(&event_id)
+                });
+
+                already_seen.insert(event_id);
+
+                if duplicated {
+                    Decoration::Duplicated(event)
+                } else {
+                    Decoration::Unique(event)
+                }
+            } else {
+                already_seen.insert(event_id);
+
+                // Bloom filter has no false negatives. We are sure the event is NOT present: we
+                // can keep it in the iterator.
+                Decoration::Unique(event)
+            }
+        })
+    }
+}
+
+/// Information about the scanned collection of events.
+#[derive(Debug)]
+pub enum Decoration<I> {
+    /// This event is not duplicated.
+    Unique(I),
+
+    /// This event is duplicated.
+    Duplicated(I),
+
+    /// This event is invalid (i.e. not well formed).
+    Invalid(I),
+}
+
+#[cfg(test)]
+mod tests {
+    use assert_matches2::assert_let;
+    use matrix_sdk_base::deserialized_responses::SyncTimelineEvent;
+    use ruma::{owned_event_id, user_id, EventId};
+
+    use super::*;
+    use crate::test_utils::events::EventFactory;
+
+    fn sync_timeline_event(event_id: &EventId) -> SyncTimelineEvent {
+        EventFactory::new()
+            .text_msg("")
+            .sender(user_id!("@mnt_io:matrix.org"))
+            .event_id(event_id)
+            .into_sync()
+    }
+
+    #[test]
+    fn test_filter_no_duplicate() {
+        let event_id_0 = owned_event_id!("$ev0");
+        let event_id_1 = owned_event_id!("$ev1");
+        let event_id_2 = owned_event_id!("$ev2");
+
+        let event_0 = sync_timeline_event(&event_id_0);
+        let event_1 = sync_timeline_event(&event_id_1);
+        let event_2 = sync_timeline_event(&event_id_2);
```

Comment:
Time for an `EventFactory 101` :nerd_face: You can curry some parameters at the factory level, to avoid dull repetition, which would justify *not* having to have the `sync_timeline_event` helper:

```suggestion
        let f = EventFactory::new().sender(user_id!("@mnt_io:matrix.org"));
        let event_0 = f.text_msg("").event_id(&event_id_0).into_sync();
        let event_1 = f.text_msg("").event_id(&event_id_1).into_sync();
        let event_2 = f.text_msg("").event_id(&event_id_2).into_sync();
```

Reply:
Thanks for the tips.

I agree it's nice but in this case it adds a non-negligible cognitive effort to read `.text_msg("")` and `.into_sync();` here. I think it's fine to keep `sync_timeline_event` as it makes the test code easier to understand, and we can focus on the real things we want to test rather than the prelude.